From 7825b40429505a921c08cb8fde76460cbc2a2202 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:11:00 -0400
Subject: [usb] pl2303: only wake up MSR queue on changes

Message-id: <1410887463-20674-589-git-send-email-dzickus@redhat.com>
Patchwork-id: 95655
O-Subject: [RHEL7 PATCH 588/591] USB: pl2303: only wake up MSR queue on changes
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit dbfd2866ac36bab52a4f37384e935eb5df76ad60
Author: Johan Hovold <jhovold@gmail.com>
Date:   Thu Jan 2 22:49:22 2014 +0100

    USB: pl2303: only wake up MSR queue on changes

    Only wake up MSR wait queue on actual modem-status changes.

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/pl2303.c b/drivers/usb/serial/pl2303.c
index eada524..acdaae4 100644
--- a/drivers/usb/serial/pl2303.c
+++ b/drivers/usb/serial/pl2303.c
@@ -120,6 +120,7 @@ MODULE_DEVICE_TABLE(usb, id_table);
 #define VENDOR_READ_REQUEST  0x01
 
 #define UART_STATE_INDEX  8
+#define UART_STATE_MSR_MASK  0x8b
 #define UART_STATE_TRANSIENT_MASK 0x74
 #define UART_DCD   0x01
 #define UART_DSR   0x02
@@ -832,15 +833,17 @@ static void pl2303_update_line_status(struct usb_serial_port *port,
  if (status & UART_BREAK_ERROR)
   usb_serial_handle_break(port);
 
- wake_up_interruptible(&port->port.delta_msr_wait);
-
- if (delta & UART_DCD) {
-  tty = tty_port_tty_get(&port->port);
-  if (tty) {
-   usb_serial_handle_dcd_change(port, tty,
+ if (delta & UART_STATE_MSR_MASK) {
+  if (delta & UART_DCD) {
+   tty = tty_port_tty_get(&port->port);
+   if (tty) {
+    usb_serial_handle_dcd_change(port, tty,
        status & UART_DCD);
-   tty_kref_put(tty);
+    tty_kref_put(tty);
+   }
   }
+
+  wake_up_interruptible(&port->port.delta_msr_wait);
  }
 }
 
-- 
1.7.1